go/private/binary.bzl

```diff
@@ -175,18 +168,24 @@ def emit_go_link_action(ctx, transitive_go_library_paths, transitive_go_librarie
   extldflags = c_linker_options(ctx) + [
       "-Wl,-rpath,$ORIGIN/" + ("../" * pkg_depth),
   ]
-  for d in cgo_deps:
-    if d.basename.endswith('.so'):
-      short_dir = d.dirname[len(d.root.path):]
-      extldflags += ["-Wl,-rpath,$ORIGIN/" + ("../" * pkg_depth) + short_dir]
 
   gc_linkopts, extldflags = _extract_extldflags(gc_linkopts, extldflags)
 
   link_opts = [
       "-L", "."
   ]
-  for path in transitive_go_library_paths:
-    link_opts += ["-L", path]
+  libs = depset()
+  cgo_deps = depset()
+  for golib in depset([library]) + library.transitive:
+    libs += [getattr(golib, "library" + mode)]
+    link_opts += ["-L", getattr(golib, "searchpath" + mode)]
```

Comment:
This feels under-abstracted to me. Perhaps `mode` could be a symbol enum and we could either turn it into a property name suffix inside this function or just do the branch. That would also let us add `"-race"` to `link_opts` implicitly, rather than relying on the caller to pass it in.